File: benchmark/synduce/constraints/ensures/maxsegstrip_full_noe.f
Config SampleSize = 20;

Inductive List = elt Int | cons {Int, List};
Inductive NList = line List | ncons {List, NList};
Inductive CNList = sglt List | cat {CNList, CNList};

cton = fix (
  \f: CNList -> NList.
  let dec = fix (
    \g: CNList -> CNList -> NList. \l: CNList. \c: CNList.
    match c with
      sglt x -> ncons {x, f l}
    | cat {x, y} -> g (cat {y, l}) x
    end
  ) in \c: CNList.
  match c with
    sglt x -> line x
  | cat {x, y} -> dec y x
  end
);

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> + h (f t)
  end
);

max = \a: Int. \b: Int. if < a b then b else a;

spec = \xs: NList.
let res = fix (
  \f: NList -> {Int, Int, Int, Int}. \xs: NList.
  match xs with
    line a ->
      let s = sum a in
        let ms = max s 0 in
          {ms, ms, ms, s}
  | ncons {h, t} ->
    let hsum = sum h in
      let res = f t in
        {
        max (+ res.1 hsum) 0,
        max res.2 (+ res.1 hsum),
        max (+ res.4 hsum) res.3,
        + res.4 hsum
        }
  end
) xs in {res.1, res.2, res.3};

target = fix (
  \f: CNList -> Compress CNList.
  let list_repr = fix (
    \g: List -> Compress List. \xs: List.
    match xs with
      elt x -> elt x
    | cons {h, t} -> cons {h, g t}
    end
  ) in \c: CNList.
  match c with
    sglt x -> sglt (list_repr x)
  | cat {l, r} -> cat {f l, f r}
  end
);

main = \c: CNList. spec (cton (target c));

File: ml-benchmark/synduce/constraints/ensures/maxsegstrip_full_noe.ml
let sampleSize = 20

type llist = Elt of int | Cons of int * llist
type nlist = Line of llist | Ncons of llist * nlist
type cnlist = Sglt of llist | Cat of cnlist * cnlist

let rec cton c =
  let rec dec l c =
    match c with
    | Sglt x -> Ncons (x, cton l)
    | Cat (x, y) -> dec (Cat (y, l)) x
  in
  match c with
  | Sglt x -> Line x
  | Cat (x, y) -> dec y x

let rec sum xs =
  match xs with
  | Elt x -> x
  | Cons (h, t) -> h + sum t

let max a b = if a < b then b else a

let spec xs =
  let rec res_f xs =
    match xs with
    | Line a ->
        let s = sum a in
        let ms = max s 0 in
        (ms, ms, ms, s)
    | Ncons (h, t) ->
        let hsum = sum h in
        let res = res_f t in
        match res with
        | (r1, r2, r3, r4) ->
            let a1 = max (r1 + hsum) 0 in
            let a2 = max r2 (r1 + hsum) in
            let a3 = max (r4 + hsum) r3 in
            let a4 = r4 + hsum in
            (a1, a2, a3, a4)
  in
  let res = res_f xs in
  match res with
  | (r1, r2, r3, _) -> (r1, r2, r3)

let rec target c =
  let rec list_repr xs =
    match xs with
    | Elt x -> Elt x
    | Cons (h, t) -> Cons (h, list_repr t)
  in
  match c with
  | Sglt x -> Sglt (list_repr x)
  | Cat (l, r) -> Cat (target l, target r)

let program c = spec (cton (target c))